server/src/main/java/org/apache/cassandra/sidecar/handlers/livemigration/LiveMigrationFileStreamHandler.java

```diff
@@ -0,0 +1,243 @@
+        if (null == dirType || dirType.isEmpty() || null == LiveMigrationDirType.find(dirType))
+        {
+            throw new HttpException(HttpResponseStatus.BAD_REQUEST.code(), "Invalid directory type: " + dirType);
+        }
+
+        String dirIndex = context.pathParam(DIR_INDEX_PARAM);
+        int index;
+        try
+        {
+            index = Integer.parseInt(dirIndex);
+        }
+        catch (NumberFormatException formatException)
+        {
+            throw new HttpException(HttpResponseStatus.BAD_REQUEST.code(), "Invalid directoryIndex: " + dirIndex);
+        }
+        if (index < 0)
+        {
+            throw new HttpException(HttpResponseStatus.BAD_REQUEST.code(), "Invalid directoryIndex: " + dirIndex);
+        }
+
+        // Path params are not used further, hence returning null.
+        return null;
+    }
+
+    @Override
+    protected void handleInternal(RoutingContext rc,
+                                  HttpServerRequest httpRequest,
+                                  @NotNull String host,
+                                  SocketAddress remoteAddress,
+                                  @Nullable Void request)
+    {
+        String reqPath = URLDecoder.decode(rc.request().path(), StandardCharsets.UTF_8);
+
+        if (reqPath.contains("/../") || reqPath.endsWith("/.."))
+        {
+            LOGGER.warn("Tried to access file using relative path({}). Rejecting the request.", reqPath);
+            rc.response().setStatusCode(HttpResponseStatus.BAD_REQUEST.code()).end();
+            return;
+        }
+
+        InstanceMetadata instanceMeta = metadataFetcher.instance(host);
+        String normalizedPath = rc.normalizedPath();
+        String localFile;
+
+        try
+        {
+            localFile = LiveMigrationInstanceMetadataUtil.localPath(normalizedPath, instanceMeta);
+        }
+        catch (IllegalArgumentException e)
+        {
+            LOGGER.warn("Invalid path", e);
+            rc.response().setStatusCode(HttpResponseStatus.NOT_FOUND.code()).end();
+            return;
+        }
+
+        Path path = Paths.get(localFile);
+
+        if (!Files.exists(path))
+        {
+            LOGGER.info("File {} not found.", localFile);
+            rc.response().setStatusCode(HttpResponseStatus.NOT_FOUND.code()).end();
+        }
+        if (Files.isDirectory(path))
+        {
+            LOGGER.info("Cannot transfer directory for request path: {}.", reqPath);
+            rc.response().setStatusCode(HttpResponseStatus.BAD_REQUEST.code()).end();
+            return;
+        }
+        if (isExcluded(path, instanceMeta))
+        {
+            LOGGER.debug("Requested file {} or one of its parent directories is excluded from Live Migration.", reqPath);
+            rc.response().setStatusCode(HttpResponseStatus.NOT_FOUND.code()).end();
+            return;
+        }
+
+        rc.put(FileStreamHandler.FILE_PATH_CONTEXT_KEY, localFile);
+        rc.next();
```

Comment:
nit: dispatch the checks to `executorPools.service()`? I am not super concerned about the I/O access, but the `isExcluded` check seems heavy. 
On success, call `rc.next()`. On failure, do nothing, as `rc` should be ended already.